src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,340 @@
-
-    return false;
+    matrix.shed_col(matrix.n_cols - 1);
+    ReshapeTimeseries(matrix, opts);
   }
-  else
-    Log::Info << "Size is " << (transpose ? matrix.n_cols : matrix.n_rows)
-        << " x " << (transpose ? matrix.n_rows : matrix.n_cols) << ".\n";
-
-  // Now transpose the matrix, if necessary.
-  if (transpose)
+  else if (opts.TimestampCol() && !opts.LabelCol())
   {
-    success = inplace_transpose(matrix, fatal);
+    matrix.shed_col(0);
+    ReshapeTimeseries(matrix, opts);
   }
 
-  Timer::Stop("loading_data");
-
-  // Finally, return the success indicator.
+   // Re-Transpose back to faciliate the cases in this function.
+  if (!opts.NoTranspose())
+  {
+    inplace_trans(matrix);
+  }
   return success;
+} 
+
+/*
+ * This function convert timeseries data collected from sensors into specific
+ * Windows that are suitable to training for machine learning.
+ * 
+ * This function is internal and should not be documented or used by users.
+ *
+ * This function only works on non-transposed matrices
+ */
+template<typename eT>
+void ReshapeTimeseries(arma::Mat<eT>& matrix,
+                       CSVOptions& opts)
+{
+  arma::Mat<eT> tmpMat = matrix.as_row();
+  size_t totalWindows = tmpMat.n_elem / opts.WindowSize();
+  //! This operation will give us by default a transposed operation.
+  //! Therefore we need to transpose back.
+  matrix = reshape(tmpMat, opts.WindowSize(), totalWindows);
+  inplace_trans(matrix);
 }
 
-// Load with mappings.  Unfortunately we have to implement this ourselves.
-template<typename eT, typename PolicyType>
+// This is the function that the user is supposed to call.
+// Other functions of this class should be labelled as private.
+template<typename MatType, typename DataOptionsType>
 bool Load(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          DatasetMapper<PolicyType>& info,
-          const bool fatal,
-          const bool transpose)
+          MatType& matrix,
+          DataOptionsType& opts)
 {
-  // Get the extension and load as necessary.
   Timer::Start("loading_data");
 
-  // Get the extension.
-  std::string extension = Extension(filename);
-
-  // Catch nonexistent files by opening the stream ourselves.
+  bool success = false;
   std::fstream stream;
-  stream.open(filename.c_str(), std::fstream::in);
-
-  if (!stream.is_open())
+  success = OpenFile(filename, opts, true, stream);
+  if (!success)
   {
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
+    return false;
+  }
 
+  success = DetectFileType(filename, opts, true, &stream);
+  if (!success)
+  {
+    Timer::Stop("loading_data");
     return false;
   }
 
-  if (extension == "csv" || extension == "tsv" || extension == "txt")
+  if (std::is_same_v<DataOptionsType, CSVOptions>)
```

Comment:
Here's another way you could structure this to use the internal detected file type in the `opts` object:

```
if (opts.FileFormat() == FileType::CSV_ASCII)
{
  // You could add extra explicit conversion constructors that throw errors when types are wrong...
  const CSVOptions csvOpts(opts); // explicitly convert to CSV options

  if (sparse) { ... }
  else if (categorical) { ... }
}
```